### ISSUE TITLE
Better `sanitizeName`

### ISSUE THREAD

**Issue by arkanoid87** [PMunch/futhark#87]
At the moment, [Futhark applies `renameCallback` before other fixed renaming rules](https://github.com/PMunch/futhark/blob/442968d1bc2c34568202e8d76a0e0a9868061059/src/futhark.nim#L181), including [nimIdentNormalize](https://nim-lang.org/docs/strutils.html#nimIdentNormalize%2Cstring) pass, which nullify all the [NEP1](
https://nim-lang.org/docs/nep1.html#style-guidelines-naming-conventions) transformations user supplied `renameCallback` apply.

I'd say `nimIdentNormalize` is not really NEP1 friendly

I've recently pushed PR to `anycase` package to support compile time string transformations https://github.com/epszaw/anycase/issues/6

What about replacing `nimIdentNormalize` in `sanitizeName` with something that is more `renameCallback` and NEP1 friendly? Or maybe provide an additional `renameTailCallback` to fix things after the fixed transformations?

**Comment by PMunch** [comment]
This code has a slight issue, it won't properly handle name collisions. For anyone who didn't follow the IRC discussion when this came up the reason why `sanitizeName` calls `nimIdentNormalize` is so that the hash table lookup of names is guaranteed to get identifiers which will collide in Nim (but which might not collide in C because it's case and underscore sensitive). The proper workaround for this is to still use the normalized name in these data-structures but change to storing the original (or `anycase`d) names and normalizing them when doing lookups.

As a motivating example I wrote a test case in the `normalize` branch which shows off the kind of thing the current solution was meant to handle. It tries to wrap this perfectly valid (albeit a bit extreme) C header file:
```c
#define my_var 1
#define myVar 2
#define myvar 3
#define MYVAR 4
#define MY_VAR 5
#define MyVar 6
#define My_Var 7
```

The current version wraps this fine and renames fields to avoid collisions as we can see in the output:
```
/home/peter/Projects/futhark/src/futhark.nim(190, 5) Hint: Renaming "myVar" to "myvarconst" [User]
/home/peter/Projects/futhark/src/futhark.nim(190, 5) Hint: Renaming "myvar" to "myvarconst00000000C690172C" [User]
/home/peter/Projects/futhark/src/futhark.nim(190, 5) Hint: Renaming "MY_VAR" to "Myvarconst" [User]
/home/peter/Projects/futhark/src/futhark.nim(190, 5) Hint: Renaming "MyVar" to "Myvarconst000000002E4AA817" [User]
/home/peter/Projects/futhark/src/futhark.nim(190, 5) Hint: Renaming "My_Var" to "Myvarconst00000000038D4D97" [User]
```
The names get appended some information before back to a hash of the original identifier.

With your version of `sanitizeName` we instead see this output:
```
/home/peter/Projects/futhark/src/futhark.nim(197, 5) Hint: Renaming "myVar" to "MyVarconst" [User]
/home/peter/Projects/futhark/src/futhark.nim(197, 5) Hint: Renaming "MY_VAR" to "MyVarconst00000000AEC77163" [User]
/home/peter/Projects/futhark/src/futhark.nim(197, 5) Hint: Renaming "MyVar" to "MyVarconst000000002E4AA817" [User]
/home/peter/Projects/futhark/src/futhark.nim(197, 5) Hint: Renaming "My_Var" to "MyVarconst00000000038D4D97" [User]
/home/peter/Projects/futhark/src/futhark.nim(908, 3) Hint: Caching Futhark output in /home/peter/.cache/nim/tnormalize_d/futhark_33DDD4433451E496.nim [User]
/home/peter/Projects/futhark/src/futhark.nim(141, 21) Hint: Declaration of Myvar already exists, not redeclaring [User]
/home/peter/Projects/futhark/src/futhark.nim(141, 21) Hint: Declaration of MYVAR already exists, not redeclaring [User]
```

As we can see it doesn't properly rename all the identifiers, and some of them are aliased together in the wrapper making `#define myvar 3` and `#define MYVAR 4` now point to `#define my_var 1`. This means that these values are not possible to get with the wrapper, and with `--nodeclguards` the code fails to compile at all because of the re-definition.

**Comment by PMunch** [comment]
Futhark no longer performs the `nimIdentNormalize` step on the output, so the original names from the C sources should be preserved, and you should now be able to create a `renameCallback` which restyles identifiers as you seem fit. Please try this out and tell me how it works!

**Comment by PMunch** [comment]
Yes I've seen something similar while wrapping Gtk. It generates the file properly, but for some reason the macro output is invalid. So if you run it twice so it grabs the cached file then everything works fine
